Project/Gem/Source/Vision/IdealVisionSystem.cpp

```diff
@@ -0,0 +1,305 @@
+#include <AzFramework/Physics/SystemBus.h>
+#include <ROS2/ROS2Bus.h>
+#include <ROS2/Sensor/ROS2SensorComponent.h>
+#include <ROS2/Utilities/ROS2Conversions.h>
+#include <ROS2/Utilities/ROS2Names.h>
+
+namespace ROS2::Demo
+{
+    constexpr char Detection2DType[] = "vision_msgs::msg::Detection2D";
+    constexpr char Detection3DType[] = "vision_msgs::msg::Detection3D";
+    constexpr char DetectionPoseArrayType[] = "geometry_msgs::msg::PoseArray";
+
+    IdealVisionSystem::IdealVisionSystem()
+    {
+        TopicConfiguration tc;
+
+        tc.m_type = DetectionPoseArrayType;
+        tc.m_topic = "detectionsPoseArray";
+        m_sensorConfiguration.m_publishersConfigurations.insert(AZStd::make_pair(tc.m_type, tc));
+
+        tc.m_type = Detection2DType;
+        tc.m_topic = "detections2D";
+        m_sensorConfiguration.m_publishersConfigurations.insert(AZStd::make_pair(tc.m_type, tc));
+
+        tc.m_type = Detection3DType;
+        tc.m_topic = "detections3D";
+        m_sensorConfiguration.m_publishersConfigurations.insert(AZStd::make_pair(tc.m_type, tc));
+
+        m_sensorConfiguration.m_frequency = 10;
+    }
+
+    AZ::Vector3 GetPointIn3D(const AZ::Vector2& point2D, float depth, AZ::Matrix3x3 cameraMatrix)
+    {
+        cameraMatrix = cameraMatrix.GetInverseFull();
+        AZ::Vector3 point3D;
+        point3D.SetX(point2D.GetX());
+        point3D.SetY(point2D.GetY());
+        point3D.SetZ(1.0f);
+        point3D = depth * cameraMatrix * point3D;
+        return point3D;
+    }
+
+    AZ::Vector2 GetPointIn2D(const AZ::Vector3& point3D, const AZ::Matrix3x3& cameraMatrix)
+    {
+        AZ::Vector3 point2D;
+        point2D = cameraMatrix * point3D;
+        point2D = point2D / point2D.GetZ();
+        return AZ::Vector2(aznumeric_cast<float>(point2D.GetX()), aznumeric_cast<float>(point2D.GetY()));
+    }
+
+    void IdealVisionSystem::Activate()
+    {
+        constexpr float MaximalDetection = 2.0f;
+
+        ROS2::CameraCalibrationRequestBus::EventResult(m_cameraMatrix, GetEntityId(), &ROS2::CameraCalibrationRequest::GetCameraMatrix);
+        ROS2::CameraCalibrationRequestBus::EventResult(m_cameraWidth, GetEntityId(), &ROS2::CameraCalibrationRequest::GetWidth);
+        ROS2::CameraCalibrationRequestBus::EventResult(m_cameraHeight, GetEntityId(), &ROS2::CameraCalibrationRequest::GetHeight);
+
+        m_frustrumPoints.clear();
+        m_frustrumPoints.push_back(AZ::Vector3(0.0f, 0.0f, 0.0f));
+        m_frustrumPoints.push_back(GetPointIn3D(AZ::Vector2(0.0f, 0.0f), MaximalDetection, m_cameraMatrix));
+        m_frustrumPoints.push_back(GetPointIn3D(AZ::Vector2(0.0f, m_cameraHeight), MaximalDetection, m_cameraMatrix));
+        m_frustrumPoints.push_back(GetPointIn3D(AZ::Vector2(m_cameraWidth, m_cameraHeight), MaximalDetection, m_cameraMatrix));
+        m_frustrumPoints.push_back(GetPointIn3D(AZ::Vector2(m_cameraWidth, 0.0f), MaximalDetection, m_cameraMatrix));
+
+        // cook PhysX mesh
```

Comment:
Consider using more subroutines.
This way, readers of code can really quickly understand what is happening. Consider:

```
void IdealVisionSystem::Activate()
{
  CreateFrustrumPoints();
  bool cookingSucceeded = CookFrustrumMesh();
  if (!cookingSucceeded)
  {
    //raise error
    return; // Note: no need to activate base component as the sensor is unusable
  }
  CreateDetectionPublishers();
  ROS2SensorComponent::Activate();
}

```
The reader needs only to few lines to understand what is happening in the Activate. If they need details on how these are done, they would look into respective function bodies, but overall it is much better in terms of cognitive load.